library/src/main/scala/scala/scalajs/niocharset/UTF_8.scala

```diff
@@ -160,20 +172,23 @@ private[niocharset] object UTF_8 extends Charset("UTF-8", Array(
       @inline
       @tailrec
       def loop(): CoderResult = {
+
         @inline
-        def finalize(read: Int, result: CoderResult): CoderResult = {
-          in.position(in.position() - read)
+        def fail(result: CoderResult): CoderResult = {
+          in.reset()
           result
         }
 
         if (!in.hasRemaining) {
           CoderResult.UNDERFLOW
         } else {
+          //mark input position so we can reset on multi-byte underflow
+          in.mark()
```

Comment:
I would prefer not to use `mark()`. It is unclear whether the charset decoder is allowed to mess up with the mark, since it could interfere with the user using `mark()` and `reset()` for their own purposes.

What about just `val savedPosition = in.position()` instead of `in.mark()`, and `in.position(savedPosition)` instead of `in.reset()`?

Reply:
Oh, I hadn't considered that. Your suggestion sounds good.